mm2src/coins_activation/src/prelude.rs

```diff
@@ -64,31 +64,78 @@ pub trait TryFromCoinProtocol {
+    CustomTokenError(CustomTokenError),
 }
 
 /// Determines the coin configuration and protocol information for a given coin or NFT ticker.
-/// In the case of NFT ticker, it's platform coin config will be returned.
-#[allow(clippy::result_large_err)]
 pub fn coin_conf_with_protocol<T: TryFromCoinProtocol>(
     ctx: &MmArc,
     coin: &str,
+    protocol_from_request: Option<CoinProtocol>,
 ) -> Result<(Json, T), MmError<CoinConfWithProtocolError>> {
     let conf = coin_conf(ctx, coin);
-    if conf.is_null() {
-        return MmError::err(CoinConfWithProtocolError::ConfigIsNotFound(coin.into()));
+    let is_ticker_in_config = !conf.is_null();
+
+    // For `protocol_from_request`: None = config-based activation, Some = custom token activation
+    match (protocol_from_request, is_ticker_in_config) {
+        // Config-based activation requested with an existing configuration
+        // Proceed with parsing protocol info from config
+        (None, true) => parse_coin_protocol_from_config(conf, coin),
+        // Custom token activation requested and no matching ticker in config
+        // Proceed with custom token config creation from protocol info
+        (Some(protocol), false) => create_custom_token_config(ctx, coin, protocol),
+        // Custom token activation requested but a coin with the same ticker already exists in config
+        (Some(_), true) => Err(MmError::new(CoinConfWithProtocolError::CustomTokenError(
+            CustomTokenError::DuplicateTickerInConfig {
+                ticker_in_config: coin.to_string(),
+            },
+        ))),
+        // Config-based activation requested but ticker not found in config
+        (None, false) => Err(MmError::new(CoinConfWithProtocolError::ConfigIsNotFound(coin.into()))),
     }
-    let coin_protocol: CoinProtocol = json::from_value(conf["protocol"].clone()).map_to_mm(|err| {
+}
+
+fn parse_coin_protocol_from_config<T: TryFromCoinProtocol>(
+    conf: Json,
+    coin: &str,
+) -> Result<(Json, T), MmError<CoinConfWithProtocolError>> {
+    let protocol = json::from_value(conf["protocol"].clone()).map_to_mm(|err| {
         CoinConfWithProtocolError::CoinProtocolParseError {
             ticker: coin.into(),
             err,
         }
     })?;
+
     let coin_protocol =
-        T::try_from_coin_protocol(coin_protocol).mm_err(|protocol| CoinConfWithProtocolError::UnexpectedProtocol {
+        T::try_from_coin_protocol(protocol).mm_err(|p| CoinConfWithProtocolError::UnexpectedProtocol {
             ticker: coin.into(),
-            protocol,
+            protocol: json!(p),
         })?;
+
+    Ok((conf, coin_protocol))
+}
+
+fn create_custom_token_config<T: TryFromCoinProtocol>(
+    ctx: &MmArc,
+    coin: &str,
+    protocol: CoinProtocol,
+) -> Result<(Json, T), MmError<CoinConfWithProtocolError>> {
+    protocol
+        .custom_token_validations(ctx)
+        .mm_err(CoinConfWithProtocolError::CustomTokenError)?;
+
+    let conf = json!({
```

Comment:
I wonder why for custom token activation only protocol is passed (why not the whole coin json)?
We could get 'decimals' from the 'coin' json.
(BTW will we not ever allow a custom coin activation? not only tokens)

Reply:
I would say the only thing that is missing is `decimals` and it's not always set in coin configs for EVM tokens as we get it here when not set https://github.com/KomodoPlatform/komodo-defi-framework/blob/2ff5448c6071e58cea6b03e989c59b548a82f08a/mm2src/coins/eth/v2_activation.rs#L414-L425
All other fields are the same as the platform coin config.

Reply:
> BTW will we not ever allow a custom coin activation?

Only tokens for now. For custom coin activation, it's a completely different thing as we don't guarantee that they will work.